Adopt Safer CPP in BiquadFilterNode.cpp, ScrollAnimator.cpp, CDMProxy.cpp

jelee53 opened the pull request:
#### 65e001848f1e0a00d0590e503e308d3028a70482
<pre>
Adopt Safer CPP in BiquadFilterNode.cpp, ScrollAnimator.cpp, CDMProxy.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=302996">https://bugs.webkit.org/show_bug.cgi?id=302996</a>
<a href="https://rdar.apple.com/165251539">rdar://165251539</a>

Reviewed by Chris Dumez and Per Arne Vollan.

Apply <a href="https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines.">https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines.</a>

No new tests needed.
* Source/WebCore/Modules/webaudio/BiquadFilterNode.cpp:
(WebCore::BiquadFilterNode::create):
(WebCore::BiquadFilterNode::setType):
(WebCore::BiquadFilterNode::getFrequencyResponse):
* Source/WebCore/Modules/webaudio/BiquadFilterNode.h:
* Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/platform/ScrollAnimator.cpp:
(WebCore::ScrollAnimator::handleSteppedScrolling):
(WebCore::ScrollAnimator::deferWheelEventTestCompletionForReason const):
(WebCore::ScrollAnimator::removeWheelEventTestCompletionDeferralForReason const):
* Source/WebCore/platform/encryptedmedia/CDMProxy.cpp:
(WebCore::CDMProxy::instance const):
(WebCore::CDMProxy::startedWaitingForKey const):
(WebCore::CDMProxy::stoppedWaitingForKey const):
(WebCore::CDMProxy::tryWaitForKeyHandle const):
(WebCore::CDMInstanceProxy::mergeKeysFrom):
(WebCore::CDMInstanceProxy::unrefAllKeysFrom):
* Source/WebCore/platform/encryptedmedia/CDMProxy.h:
(WebCore::CDMInstanceProxy::~CDMInstanceProxy):
* Source/WebCore/platform/encryptedmedia/clearkey/CDMClearKey.cpp:
* Source/WebCore/platform/encryptedmedia/clearkey/CDMClearKey.h:
* Source/WebCore/platform/graphics/gstreamer/eme/CDMThunder.cpp:
* Source/WebCore/platform/graphics/gstreamer/eme/CDMThunder.h:

Canonical link: <a href="https://commits.webkit.org/304121@main">https://commits.webkit.org/304121@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dfe406685269a6a7c1af30f60249dcf68c9d58ce

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134359 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6832 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45575 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141901 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86358 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/c6fd61ae-0d83-4a71-b084-d53292fe1ede) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136229 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7426 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6699 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102704 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/69963 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/285c3a60-f1be-452b-b6bc-6c66ecfe21e3) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137306 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5175 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120414 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83495 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/eb50dbd7-cf29-48e0-8760-cbd31dfc1f7d) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5040 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2663 "Passed tests") | [⏳ 🛠 wpe-cairo-libwebrtc ](https://ews-build.webkit.org/#/builders/WPE-Cairo-LibWebRTC-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/114246 "Found 1 new API test failure: TestWebKitAPI.ProcessSwap.PageOverlayLayerPersistence (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38555 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144583 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6512 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39138 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111102 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6588 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5473 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111369 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4876 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116690 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60294 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20789 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6564 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34890 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6389 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70129 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6624 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6501 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/webaudio/BiquadFilterNode.cpp:
```diff
@@ -80,7 +80,7 @@ ExceptionOr<void> BiquadFilterNode::getFrequencyResponse(const Ref<Float32Array>
 
     if (length)
-        biquadProcessor()->getFrequencyResponse(length, frequencyHz->typedSpan(), magResponse->typedMutableSpan(), phaseResponse->typedMutableSpan());
+        CheckedPtr { biquadProcessor() }->getFrequencyResponse(length, frequencyHz->typedSpan(), magResponse->typedMutableSpan(), phaseResponse->typedMutableSpan());
```
I recommend adding a `checkedBiquadProcessor()` getter instead since this is needed more than once.

jelee53 replied:
I did make this helper function but I didn't end up making one for the other file, CDMProxy (aka `CheckedPtr { m_instance }` in `CDMProxy::stoppedWaitingForKey`  and startedWaitingForKey )

 The pre-existing getter `const CDMInstanceProxy* instance() const` is a const member function and `startedWaitingForKey ` and `stoppedWaitingForKey` both alter the object.

cdumez commented on Source/WebCore/platform/ScrollAnimator.cpp:
```diff
@@ -409,18 +409,14 @@ void ScrollAnimator::stopAnimationCallback(ScrollingEffectsController&)
-        return;
-
-    m_wheelEventTestMonitor->deferForReason(identifier, reason);
+    if (RefPtr wheelEventTestMonitor = m_wheelEventTestMonitor)
```
We can even go with `monitor` for the variable name given the small scope.

cdumez commented on Source/WebCore/platform/ScrollAnimator.cpp:
```diff
@@ -409,18 +409,14 @@ void ScrollAnimator::stopAnimationCallback(ScrollingEffectsController&)
-        return;
-    
-    m_wheelEventTestMonitor->removeDeferralForReason(identifier, reason);
+    if (RefPtr wheelEventTestMonitor = m_wheelEventTestMonitor)
```
ditto.

cdumez commented on Source/WebCore/platform/encryptedmedia/CDMProxy.cpp:
```diff
@@ -184,15 +184,15 @@ void CDMProxy::startedWaitingForKey() const
     LOG(EME, "EME - CDMProxy - started waiting for a key");
     ASSERT(m_instance);
-    m_instance->startedWaitingForKey();
+    RefPtr { m_instance }->startedWaitingForKey();
```
Given the locking above, I'm very concerned this may be multi-threaded code and `m_instance` is using a non-thread-safe ref count. We may have to use `CheckedPtr` and `CanMakeThreadSafeCheckedPtr<>` instead to make static analysis happy.

jelee53 replied:
Yes, indeed we are failing `wpe-wk2` because m_instance is not thread safe. I will fix CDMInstance to be a CheckedPtr instead.

jelee53 commented on Source/WebCore/platform/encryptedmedia/CDMProxy.h:
```diff
@@ -306,15 +307,17 @@ class CDMInstanceProxy;
 
 private:
-    WeakPtr<CDMInstanceProxy> m_instance;
+    CheckedPtr<CDMInstanceProxy> m_instance;
```
TODO: I'll change this back to WeakPtr when I'm back from vacation.

jelee53 replied:
I should also probably add the destructors for the child classes, too.

pvollan commented on Source/WebCore/platform/encryptedmedia/CDMProxy.cpp:
```diff
@@ -207,15 +209,15 @@ std::optional<Ref<KeyHandle>> CDMProxy::tryWaitForKeyHandle(const KeyIDType& key
         Locker locker { m_keysLock };
 
-        m_keysCondition.waitFor(m_keysLock, CDMProxy::MaxKeyWaitTimeSeconds, [this, keyID, client = WTFMove(client), &wasKeyAvailable]() {
+        m_keysCondition.waitFor(m_keysLock, CDMProxy::MaxKeyWaitTimeSeconds, [this, protectedThis = Ref { *this }, keyID, client = WTFMove(client), &wasKeyAvailable]() {
```
Perhaps we can avoid capturing `this` now that we capture `protectedThis`?

jelee53 replied:
Discussed offline with Per and we can leave this as-is.

jelee53 committed:
Adopt Safer CPP in BiquadFilterNode.cpp, ScrollAnimator.cpp, CDMProxy.cpp
https://bugs.webkit.org/show_bug.cgi?id=302996
rdar://165251539

Reviewed by Chris Dumez and Per Arne Vollan.

Apply https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines.

No new tests needed.
* Source/WebCore/Modules/webaudio/BiquadFilterNode.cpp:
(WebCore::BiquadFilterNode::create):
(WebCore::BiquadFilterNode::setType):
(WebCore::BiquadFilterNode::getFrequencyResponse):
* Source/WebCore/Modules/webaudio/BiquadFilterNode.h:
* Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/platform/ScrollAnimator.cpp:
(WebCore::ScrollAnimator::handleSteppedScrolling):
(WebCore::ScrollAnimator::deferWheelEventTestCompletionForReason const):
(WebCore::ScrollAnimator::removeWheelEventTestCompletionDeferralForReason const):
* Source/WebCore/platform/encryptedmedia/CDMProxy.cpp:
(WebCore::CDMProxy::instance const):
(WebCore::CDMProxy::startedWaitingForKey const):
(WebCore::CDMProxy::stoppedWaitingForKey const):
(WebCore::CDMProxy::tryWaitForKeyHandle const):
(WebCore::CDMInstanceProxy::mergeKeysFrom):
(WebCore::CDMInstanceProxy::unrefAllKeysFrom):
* Source/WebCore/platform/encryptedmedia/CDMProxy.h:
(WebCore::CDMInstanceProxy::~CDMInstanceProxy):
* Source/WebCore/platform/encryptedmedia/clearkey/CDMClearKey.cpp:
* Source/WebCore/platform/encryptedmedia/clearkey/CDMClearKey.h:
* Source/WebCore/platform/graphics/gstreamer/eme/CDMThunder.cpp:
* Source/WebCore/platform/graphics/gstreamer/eme/CDMThunder.h:

Canonical link: https://commits.webkit.org/304121@main
